File: 206. Reverse Linked List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode * dummy = new ListNode(-1);
        dummy->next = head;
        ListNode * p1 = head;
        ListNode * p2 = head;
        ListNode * nextnode = NULL;
        
        while (p1 != NULL)
        {
            p2 = p1->next;
            p1->next = nextnode;
            dummy->next = p1;
            nextnode = p1;
            p1 = p2;
        }
        return dummy->next;
    }
};

File: 206. Reverse Linked List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        dummy = ListNode(-1)
        dummy.next = head
        p1 = head
        p2 = head
        nextnode = None
        
        while p1:
            p2 = p1.next
            p1.next = nextnode
            dummy.next = p1
            nextnode = dummy.next
            p1 = p2
        return dummy.next
